include/net/mqtt.h

```diff
@@ -0,0 +1,629 @@
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ *
+ * @note Shall be called before initiating any procedures on the module.
+ * @note If module initialization fails, no module APIs shall be called.
+ */
+int mqtt_init(void);
+
+/**
+ * @brief Initializes the client instance.
+ *
+ * @param[in] client Client instance for which the procedure is requested.
+ *                   Shall not be NULL.
+ *
+ * @note Shall be called before connecting the client in order to avoid
+ *       unexpected behavior caused by uninitialized parameters.
+ */
+void mqtt_client_init(struct mqtt_client *const client);
+
+/**
+ * @brief API to request new MQTT client connection.
+ *
+ * @param[out] client Client instance for which the procedure is requested.
+ *                    Shall not be NULL.
+ *
+ * @note This memory is assumed to be resident until mqtt_disconnect is called.
+ * @note Any subsequent changes to parameters like broker address, user name,
+ *       device id, etc. have no effect once MQTT connection is established.
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ *
+ * @note Default protocol revision used for connection request is 3.1.0. Please
+ *       set client.protocol_version = MQTT_VERSION_3_1_1 to use protocol 3.1.1.
+ * @note If more than one simultaneous client connections are needed, please
+ *       define MQTT_MAX_CLIENTS to override default of 1.
+ * @note Please define MQTT_KEEPALIVE time to override default of 1 minute.
+ * @note Please define MQTT_MAX_PACKET_LENGTH time to override default of
+ *       128 bytes. Ensure the system has enough memory for the new length
+ *       per client.
+ */
+int mqtt_connect(struct mqtt_client *const client);
+
+/**
+ * @brief API to publish messages on topics.
+ *
+ * @param[in] client Client instance for which the procedure is requested.
+ *                   Shall not be NULL.
+ * @param[in] param Parameters to be used for the publish message.
+ *                  Shall not be NULL.
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ *
+ * @note Default protocol revision used for connection request is 3.1.0. Please
+ *       set client.protocol_version = MQTT_VERSION_3_1_1 to use protocol 3.1.1.
+ */
+int mqtt_publish(struct mqtt_client *const client,
+		 struct mqtt_publish_param const *const param);
+
+/**
+ * @brief API used by subscribing client to send acknowledgment to the broker.
+ *        Applicable only to QoS 1 publish messages.
+ *
+ * @param[in] client Client instance for which the procedure is requested.
+ *                   Shall not be NULL.
+ * @param[in] param Identifies message being acknowledged.
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ *
+ * @note Default protocol revision used for connection request is 3.1.0. Please
+ *       set client.protocol_version = MQTT_VERSION_3_1_1 to use protocol 3.1.1.
+ */
+int mqtt_publish_ack(struct mqtt_client *const client,
```

Comment:
Also, could go user-friendly and call it e.g. mqtt_publish_qos1_ack(), likewise for qos2.

Reply:
Not sure about it, if we had for instance QoS1 and QoS2 ack messages, but here ack is QoS1-bound, same with others.

Reply:
> but here ack is QoS1-bound, same with others

I'm not sure I understood the response. If here ack is QoS1-bound, then we can add that fact to the function name, and that's what I propose.